PyInstaller/building/osx.py

```diff
@@ -123,6 +129,416 @@ def _check_guts(self, data, last_build):
+        # `QtQuick.2`, `QtQuick/Controls.2`, `QtQuick/Particles.2`, `QtQuick/Templates.2`, etc.
+        #
+        # In this step, we only prepare symlink entries that link the original directory name (with dot) to the modified
+        # one (with dot replaced). The parent paths for collected files are modified in later step(s).
+        for directory_path, directory_type in directory_types.items():
+            # .framework bundle directories contain a dot in the name, but are allowed that.
+            if directory_type == _FRAMEWORK_DIR_TYPE:
+                continue
+
+            # Data-only directories are fully located in `Contents/Resources` and cross-linked to `Contents/Frameworks`
+            # at directory level, so they are also allowed a dot in their name.
+            if directory_type == _DATA_DIR_TYPE:
+                continue
+
+            # Apply the work-around, if necessary...
+            if '.' not in directory_path.name:
+                continue
+
+            logger.debug(
+                "Creating symlink to work around the dot in the name of directory %r (%s)...", str(directory_path),
+                directory_type
+            )
+
+            # Create a SYMLINK entry, but only for this level. In case of nested directories with dots in names, the
+            # symlinks for ancestors will be created by corresponding loop iteration.
+            bundle_toc.append((
+                os.path.join('Contents/Frameworks', directory_path),
+                directory_path.name.replace('.', DOT_REPLACEMENT),
+                'SYMLINK',
+            ))
+
+        # Step 4: process the entries for collected files, and decide whether they should be placed into
+        # `Contents/MacOS`, `Contents/Frameworks`, or `Contents/Resources`, and whether they should be cross-linked into
+        # other directories.
+        for orig_dest_name, src_name, typecode in toc:
+            orig_dest_path = pathlib.PurePath(orig_dest_name)
+
+            # Special handling for EXECUTABLE and PKG entries
+            if typecode == 'EXECUTABLE':
+                # Place into `Contents/MacOS`, ...
+                file_dest = os.path.join('Contents/MacOS', orig_dest_name)
+                bundle_toc.append((file_dest, src_name, typecode))
+                # ... cross-link into `Contents/Frameworks`, ...
+                symlink_dest = os.path.join('Contents/Frameworks', orig_dest_name)
+                symlink_ref = self._compute_relative_crosslink(symlink_dest, file_dest)
+                bundle_toc.append((symlink_dest, symlink_ref, 'SYMLINK'))
+                # ... and cross-link into `Contents/Resources`.
+                symlink_dest = os.path.join('Contents/Resources', orig_dest_name)
+                symlink_ref = self._compute_relative_crosslink(symlink_dest, file_dest)
+                bundle_toc.append((symlink_dest, symlink_ref, 'SYMLINK'))
```

Comment:
Is cross linking the executable really necessary? `sys.executable` (the only mechanism people should be using to find the executable) already points to the original in `Contents/MacOS` so I'd argue that anyone who utilizes these symlinks is doing it wrong. Removing these lines + #7713 would resolve #7314.

Reply:
> Is cross linking the executable really necessary? `sys.executable` (the only mechanism people should be using to find the executable) already points to the original in `Contents/MacOS` so I'd argue that anyone who utilizes these symlinks is doing it wrong. Removing these lines + #7713 would resolve #7314.

You're right - it was originally put there for the sake of completeness. But now that everything but executable has been relocated to `Contents/Frameworks` and now that #7713 is on the table as well, it would make more sense to resolve #7314 by removing those lines. I'll push a commit once the CI catches up with its backlog.